File: scripting/jvmMain/java/com/dropbear/rendering/DebugDrawNative.java
package com.dropbear.rendering;

import com.dropbear.EucalyptusCoreLoader;
import com.dropbear.math.Quaterniond;
import com.dropbear.math.Vector3d;
import com.dropbear.utils.Colour;

public class DebugDrawNative {
    static {
        new EucalyptusCoreLoader().ensureLoaded();
    }

    public static native void drawLine(long graphicsContextPtr, Vector3d start, Vector3d end, Colour colour);
    public static native void drawRay(long graphicsContextPtr, Vector3d origin, Vector3d dir, Colour colour);
    public static native void drawArrow(long graphicsContextPtr, Vector3d start, Vector3d end, Colour colour);
    public static native void drawPoint(long graphicsContextPtr, Vector3d pos, float size, Colour colour);
    public static native void drawCircle(long graphicsContextPtr, Vector3d center, float radius, Vector3d normal, Colour colour);
    public static native void drawSphere(long graphicsContextPtr, Vector3d center, float radius, Colour colour);
    public static native void drawGlobe(long graphicsContextPtr, Vector3d center, float radius, int latLines, int lonLines, Colour colour);
    public static native void drawAabb(long graphicsContextPtr, Vector3d min, Vector3d max, Colour colour);
    public static native void drawObb(long graphicsContextPtr, Vector3d center, Vector3d halfExtents, Quaterniond rotation, Colour colour);
    public static native void drawCapsule(long graphicsContextPtr, Vector3d a, Vector3d b, float radius, Colour colour);
    public static native void drawCylinder(long graphicsContextPtr, Vector3d center, float halfHeight, float radius, Vector3d axis, Colour colour);
    public static native void drawCone(long graphicsContextPtr, Vector3d apex, Vector3d dir, float angle, float length, Colour colour);
}

File: scripting/nativeMain/kotlin/com/dropbear/rendering/DebugDraw.native.kt
@file:OptIn(ExperimentalForeignApi::class)

package com.dropbear.rendering

import com.dropbear.DropbearEngine
import com.dropbear.ffi.generated.*
import com.dropbear.math.Quaterniond
import com.dropbear.math.Vector3d
import com.dropbear.utils.Colour
import kotlinx.cinterop.*

internal actual fun DebugDraw.drawLineNative(start: Vector3d, end: Vector3d, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_line(g, allocVec3(start).ptr, allocVec3(end).ptr, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawRayNative(origin: Vector3d, dir: Vector3d, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_ray(g, allocVec3(origin).ptr, allocVec3(dir).ptr, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawArrowNative(start: Vector3d, end: Vector3d, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_arrow(g, allocVec3(start).ptr, allocVec3(end).ptr, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawPointNative(pos: Vector3d, size: Float, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_point(g, allocVec3(pos).ptr, size, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawCircleNative(center: Vector3d, radius: Float, normal: Vector3d, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_circle(g, allocVec3(center).ptr, radius, allocVec3(normal).ptr, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawSphereNative(center: Vector3d, radius: Float, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_sphere(g, allocVec3(center).ptr, radius, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawGlobeNative(center: Vector3d, radius: Float, latLines: Int, lonLines: Int, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_globe(g, allocVec3(center).ptr, radius, latLines.toUInt(), lonLines.toUInt(), allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawAabbNative(min: Vector3d, max: Vector3d, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_aabb(g, allocVec3(min).ptr, allocVec3(max).ptr, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawObbNative(center: Vector3d, halfExtents: Vector3d, rotation: Quaterniond, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_obb(g, allocVec3(center).ptr, allocVec3(halfExtents).ptr, allocQuat(rotation).ptr, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawCapsuleNative(a: Vector3d, b: Vector3d, radius: Float, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_capsule(g, allocVec3(a).ptr, allocVec3(b).ptr, radius, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawCylinderNative(center: Vector3d, halfHeight: Float, radius: Float, axis: Vector3d, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_cylinder(g, allocVec3(center).ptr, halfHeight, radius, allocVec3(axis).ptr, allocColour(colour).ptr)
}

internal actual fun DebugDraw.drawConeNative(apex: Vector3d, dir: Vector3d, angle: Float, length: Float, colour: Colour) = memScoped {
    val g = DropbearEngine.native.graphicsContextHandle ?: return@memScoped
    dropbear_debug_draw_cone(g, allocVec3(apex).ptr, allocVec3(dir).ptr, angle, length, allocColour(colour).ptr)
}
